### PR TITLE
[KEP-2400] add no swap as the default option for swap

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Implementation of beta2 findings for KEP-2400.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Waiting on KEP PR.
#### Does this PR introduce a user-facing change?

```release-note
kubelet: the `.memorySwap.swapBehavior` field in kubelet configuration accepts a new value `NoSwap` and makes this the default if unspecified; the previously accepted `UnlimitedSwap` value has been dropped.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/issues/2400
```

### CODE REVIEW

**Comment by kannon92** [comment]
/hold
holding for KEP PR as this is not yet approved for the design.

**Comment by aojea** [comment]
sounds like a more elegant solution 👏

**Comment by pacoxu** [comment]
/cc
add to my todo list of v1.30

**Comment by kannon92** [comment]
/hold cancel
KEP got approved so this PR is good to go.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (harche) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:171]: Just wondering if we are removing the support for `UnlimitedSwap` and `LimitedSwap` is the only type of `SwapBehavior` supported, then does it even need to be explicitly handled like this?

File: pkg/kubelet/apis/config/validation/validation_test.go
Comment (kannon92) [reply pkg/kubelet/apis/config/validation/validation_test.go:367]: I think we have to have `NoSwap` as an option. We need a way to allow people to use `--fail-swap-on=false` but disable swap usage on the workloads.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:171]: Well we may have new options and we have `LimitedSwap` and `NoSwap`. Since `memorySwapBehavior` assumes "" or `NoSwap` I'm not sure how we can get away with not setting these. We should probably handle `LimitedSwap` and `NoSwap`. Users may either set to `NoSwap` or remove the option so we should probably support both.

File: pkg/kubelet/apis/config/validation/validation_test.go
Comment (harche) [reply pkg/kubelet/apis/config/validation/validation_test.go:367]: I agree, we need to have a way for people to use `--fail-swap-on=false`. Thanks.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (harche) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:171]: Makes sense, thanks.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (haircommander) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:192]: hm can we do this to a beta feature? We're taking a former default with a beta feature gate and then totally upending the meaning.

Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:192]: Yea :( The default we had to drop and that is why I documented in the KEP. It is a breaking change but we want to drop that support so not sure what we should do in that case.

Comment (haircommander) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:192]: https://github.com/kubernetes/enhancements/blob/e14ee585ad86ae38dd13583bedee2d25f9eadaf1/keps/sig-node/2400-node-swap/README.md#beta-2
gotcha missed that piece

**Review by harche — COMMENTED** [review]
/lgtm

**Review by mrunalp — APPROVED** [review]

File: pkg/kubelet/apis/config/types.go
Comment (iholder101) [inline pkg/kubelet/apis/config/types.go:660]: Although it's not directly related to your changes, this documentation seems to be inaccurate.
Perhaps something as follows would be better:

```suggestion
	// "LimitedSwap": workload swap usage is limited. The swap limit is proportionate to the container's memory request.
```

The limit formula (as can be seen [here](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#steps-to-calculate-swap-limit)) is:
`limit=<swap_size>*(<memory_request>/<node_memory_capacity>)`.

I can take care of this (and maybe other instances of outdated documentation?) in a separate PR if you think it's out-of-scope for this one.

File: pkg/features/kube_features.go
Comment (iholder101) [inline pkg/features/kube_features.go:1060]: :+1:

**Review by iholder101 — COMMENTED** [review]
Thank you @kannon92!

I like the idea of turning swap on by default and introducing a `NoSwap` behavior. Elegant and clean.

/lgtm

File: pkg/kubelet/apis/config/types.go
Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:660]: Good call. I'll add this.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go]: updated.

**Comment by kannon92** [comment]
/assign @liggitt

Please feel free to delegate.

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [inline pkg/kubelet/apis/config/types.go:668]: There are two ways people can use the beta swap feature in 1.29 that this change will disrupt:

1. running 1.29 kubelets with `--fail-swap-on=false --feature-gates=NodeSwap=true` and `swapBehavior: UnlimitedSwap`
   * configurations like this will fail to start in 1.30 with validation errors because we've dropped UnlimitedSwap support
   * it sounds like the feedback from the beta was that we couldn't safely offer this option
   * this is at least very clear at startup
2. running 1.29 kubelets with `--fail-swap-on=false --feature-gates=NodeSwap=true` and not specifying `swapBehavior` at all in the kubelet config
  * configurations like this will switch from running LimitedSwap to not enabling swap at all
  * this is likely to be missed by people and seems more problematic to me

I would expect people who didn't opt into swap previously to not accidentally start using it, and people who *did* opt into swap in 1.29 to not accidentally *stop* using it.

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go:1066]: I don't really understand what you mean here?

Comment (kannon92) [reply pkg/features/kube_features.go:1066]:
```suggestion
	NodeSwap: {Default: true, PreRelease: featuregate.Beta}, // default to true in 1.30
```
Is this what you mean?

Comment (liggitt) [reply pkg/features/kube_features.go:1069]:
```
	// owner: @iholder101
	// alpha: v1.22
	// beta1: v1.28. For more info, please look at the KEP: https://kep.k8s.io/2400.
	// beta2: v1.30. Default on.
	//
	// Permits kubelet to run with swap enabled
	NodeSwap featuregate.Feature = "NodeSwap"
```

Comment (kannon92) [reply pkg/features/kube_features.go:1069]: Pushed a fix. https://github.com/kubernetes/kubernetes/pull/122745/commits/20a7b03cc3b1830fc14c4f59d901b0cb34e759ca

File: pkg/kubelet/apis/config/types.go
Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: Is this something we can solve with documentation? I'm not really sure how to handle the second point to be honest.

@mrunalp @SergeyKanzhelev WDYT?

By removing UnlimitedSwap, we would only allow people to use Swap if their pods are burstable. So we expect a reduced scope of users but it does impact those who used UnlimitedSwap.

Comment (mrunalp) [reply pkg/kubelet/apis/config/types.go:668]: What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

Comment (sftim) [reply pkg/kubelet/apis/config/types.go:668]: We can also document how to verify that swap is available in your cluster. Does it show up in `.status.capacity` for the Node? (if not, I wonder if it should appear somewhere else).

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: I added a PR to nfd to add a swap label to nodes if swap exists. That of course requires NFD.

For kube, we worry about putting that in .status.capacity as only burstable pods could utilize swap and you are not guaranteed to use swap. Its only at pressure.

We did add stats to summary stats in kubelet that one could use to verify swap capacity.

Comment (sftim) [reply pkg/kubelet/apis/config/types.go:668]: >  we worry about putting that in `.status.capacity` as only burstable pods could utilize swap and you are not guaranteed to use swap

Related, I wonder if we should report swap pressure as a condition (we don't necessarily have to trigger a taint when the condition occurs).

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: Yea that is one thing we may consider in future KEP work. We decided to not support UnlimitedSwap because it isn't clear what it means to have swap pressure and what to do in that scenario. And with UnlimtedSwap you can easily use up all the swap.

@iholder101, @fabiand and I talked a bit about this actually. We are still researching what it means to have swap pressure and what kind of evictions we would want to take. Swap is highly dependent on the kernel and the type of swap (disk, memory).

Comment (sftim) [reply pkg/kubelet/apis/config/types.go:668]: As a starting point, how about setting a threshold for page file use and mapping that to a condition but taking no further action? The default can be to set the condition at 100% use, with a configuration setting to allow a lower figure.

At least then, if someone wants nodes tainted when the condition happens, they can make their own controller and supply their own taint. And of course the mechanism to evict based on taints is already well defined.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: Main thing we find with users of swap, they find that the eviction manager is getting in the way. So a few users have disabled memory evictions if swap is enabled.

https://github.com/kubernetes/enhancements/issues/2400#issuecomment-1943207418

Its not an easier answer at the moment so we took a more conservative approach with this kep and focused on a smaller use case for LimitedSwap.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: @liggitt WDYT of @mrunalp suggestion?

I drafted up the work https://github.com/kubernetes/kubernetes/pull/123645 in case we decide to go that route. I'll merge that work into here but wanted to see if this is the path we would take.

We would fail the case "" and add a validation error on kubelet startup. We default to NoSwap starting from 1.30.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: and @sftim We are keeping this feature in beta and we are open to SwapPressure but I think I'd like some more suggestions from users on what they would want.

Is there an objection to adding that as a followup?

We would like to add this functionality as this would block us on GAing eventually. We discovered that CI/users do in fact use `--fail-swap-on=false` and they don't expect swap to be used.

Comment (iholder101) [reply pkg/kubelet/apis/config/types.go:668]: > configurations like this will switch from running LimitedSwap to not enabling swap at all

A small note: in 1.29 the default behavior was actually UnlimitedSwap, which is now dropped, therefore we have no other option but to change the default. IMHO doing so for a feature that did not graduate to full Beta yet is reasonable and we should be able to get around it with documentation.

> What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

While it's true that @mrunalp's suggestion would be more visible, it's probably worse for the longer term than what's suggested in this PR since:
1. NodeSwap feature gate would be on by default. So if `--fail-swap-on=false` is provided kubelet would fail. IMO it's somewhat of an odd behavior, since the user explicitly asked kubelet not to fail when swap is on, but now he needs to take two steps to make this happen. We can instead deprecate `--fail-swap-on` and make it a no-op, but I don't really see why this is better than what's proposed by this PR.
2. This would cause (our and possibly others') CI to fail, since swap behavior is not provided.

> Yea that is one thing we may consider in future KEP work.

:+1:
I think that this KEP provides a stable, yet minimal and basic use of swap. There are many further areas to explore and expand (evictions, more customized control, better visibility, and much more), which I think should be done in a follow-up KEP.

This way we would able to provide k8s users with a stable yet basic use of swap while we can gain feedback,  inputs, and understand use-cases better in order to present more advanced swap in the future.

Comment (rphillips) [reply pkg/kubelet/apis/config/types.go:668]: Instead of a string we can create a first class type.

```
type SwapBehavior string

const (
  UnlimitedSwap = "UnlimitedSwap"
  NoSwap = "NoSwap"
)

...
```

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: I would love that but I was thinking that would be a breaking change…

we should’ve used a first class type in alpha as that would make support easier.

Comment (sftim) [reply pkg/kubelet/apis/config/types.go:668]: We prefer not to make breaking changes, but we can (the feature is not GA).

Comment (liggitt) [reply pkg/kubelet/apis/config/types.go:668]: > What if we don't default and require the swap mode to be explicitly set? If it isn't set in the 2nd case above then we can error out.

I guess requiring someone who has opted into running with swap (by changing the `--fail-swap-on=false` arg) to explicitly indicate the workload swap behavior they want is probably the least bad option here.

There's three user config knobs related to swap today:

1. `featureGates["NodeSwap"]` (or `--feature-gates=NodeSwap=...` kubelet flag)
  * defaults to false <1.30
  * desired to default to true >=1.30 (as ~all feature gates are expected to do eventually)
2. `failSwapOn` (or `--fail-swap-on` kubelet flag)
  * defaults true
3. `memorySwap.swapBehavior`
  * defaults to UnlimitedSwap <1.30
  * desired to not default to UnlimitedSwap >=1.30

If a future release wants to change `--fail-swap-on` to default to `false`, there will be another can of worms here.

> @kannon92:
> We discovered that CI/users do in fact use --fail-swap-on=false and they don't expect swap to be used.

Oof... I wonder how many of them understand implications like #105978

> @iholder101:
> A small note: in 1.29 the default behavior was actually UnlimitedSwap, which is now dropped, therefore we have no other option but to change the default. IMHO doing so for a feature that did not graduate to full Beta yet is reasonable and we should be able to get around it with documentation.

Good to know.

> @iholder101:
> We can instead deprecate --fail-swap-on and make it a no-op, but I don't really see why this is better than what's proposed by this PR.

Changing the default for `--fail-swap-on` or making it a no-op is a different discussion. I expected a safer resolution of #105978 before defaulting the kubelet to be willing to create tmpfs volumes that could actually be persisted to disk.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: Yea, we discovered this problem with trying to fail this feature for cgroupv1. We figured we could use --failSwapOn=false as who is using that field if we don't have swap enabled. Well kind jobs in our CI provision failSwap=false and are using cgroup v1. So that was fun.

We figured that defaulting NoSwap can at least not break the failSwapOn=false usecase with cgroupv1. I sorta view failSwap as separate to this feature at the moment. It was thought in the early stages of this feature that we can safely view it as a toggle for this feature but we discovered problems.

Its a prereq for this feature but it existed before this feature so we can't view it as a toggle for this feature.

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go]: updated.

File: pkg/kubelet/apis/config/types.go
Comment (fabiand) [reply pkg/kubelet/apis/config/types.go:668]: > this is likely to be missed by people and seems more problematic to me

@liggitt I do not disagree, but wasn't one goal of this beta2 to create more certainty and safety for users?
To me the decision about the new defaults (NoSwap) and removing features (UnlimitedSwap) seems to be a direct consequence of this research. Aka the goal of beta2 was reached :)

How about adding a warning or info to the logs of the relevant component in order to make this change easier to discover (thus a warning) and to notify the user about these two changes and their mitigation.
After all the mitigation is at least rather easy.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: I followed up with @liggitt on slack.

> can we detect when they opted into swap with --fail-swap-on=false and no swapBehavior config option was provided, and output a warning at least in the kubelet log, so we aren't silently switching from making workloads use swap to not use swap? 

I am adding a log to make it clear that we are using NoSwap if swapBehavior is not specified. I can't really use --fail-swap-on=false but if fail-swap-on=true is set and we are using swap then kubelet would fail to start so I think that is fine.

Comment (kannon92) [reply pkg/kubelet/apis/config/types.go:668]: Added:

`	switch m.memorySwapBehavior {
	default:
		klog.InfoS("NoSwap is set due to memorySwapBehavior not specified", "memorySwapBehavior", m.memorySwapBehavior)
		swapConfigurationHelper.ConfigureNoSwap(lcr)
	}`

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:540]:
```suggestion
	// Permits kubelet to run with swap enabled.
```

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (liggitt) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:179]: move the feature gate check to the top of the function, otherwise we'll be outputting lots of lots per container unnecessarily

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (liggitt) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:192]: log this at startup, not for every container

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:173]: I think this log is useful as is. LimitedSwap is dependent on the container/pod as we only swap for BurstableQos. I think having the logs at the container level is useful here.

Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:173]: sorry yea I pushed up a fix for this. Had to move a lot of this information due startup on kubelet app.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:168]: pushed up a fix for this.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:178]: I removed logs here.

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go]: updated.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go]: moved this to startup.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (liggitt) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:168]: it still looks like we'll be logging `klog.InfoS("No swap cgroup controller present"` for every container for the whole kubelet lifetime... that seems like log spam, right?

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/122745#discussion_r1513387636 is my only outstanding comment, the rest of the changes look reasonable to me

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (liggitt) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:168]: huh, I assumed `configureContainerSwapResources` was only called when NodeSwap feature was enabled, and defaulting it on in this PR would increase log spam, but it looks like it has already been called unconditionally, so any log spam already exists.

Given that, I'm ok with trying to mitigate the log spam in a follow-up PR (though I think that should definitely be done). We probably want to output the message once at startup very visibly.

A follow-up PR to relocate this log message to startup seems fine to me.

**Comment by liggitt** [comment]
/approve
for kubelet config API changes

/hold for last-review of the impl changes (relocating logs, tweak to validation, etc) by @mrunalp

File: pkg/kubelet/cm/container_manager_linux.go
Comment (mrunalp) [inline pkg/kubelet/cm/container_manager_linux.go]: nit (non-blocking): a -> the

Comment (mrunalp) [reply pkg/kubelet/cm/container_manager_linux.go]: Looks like we dropped this in the next commit. Can we fixup the commits so they are additive?

Comment (kannon92) [reply pkg/kubelet/cm/container_manager_linux.go]: I am not sure how to do that. I can squash all them if thats easier?

Comment (mrunalp) [reply pkg/kubelet/cm/container_manager_linux.go]: Yeah, I find other cases where we have changed code from previous commit to next, so could squash if that's easiest.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (mrunalp) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:168]: @harche @iholder101 Please follow up on this one.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (kannon92) [reply pkg/kubelet/cm/container_manager_linux.go]: Squashed and pushed up.

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux.go
Comment (kannon92) [reply pkg/kubelet/kuberuntime/kuberuntime_container_linux.go:168]: https://github.com/kubernetes/kubernetes/issues/123728

File: cmd/kubelet/app/server.go
Comment (yujuhong) [inline cmd/kubelet/app/server.go:808]: Shouldn't we return an error since this is not supported? What do we want to let kubelet continue running for an unsupported combination?

Comment (kannon92) [reply cmd/kubelet/app/server.go:808]: Thats a good point. We never failed in the cases around this swap feature so it has become difficult to enforce that from this point.

We had an issue to fail cgroupv1 and this feature but then we discovered that people were using `--fail-swap-on=false` with cgroupv1. So I may have been more pragmatic here than I needed.

Comment (yujuhong) [reply cmd/kubelet/app/server.go:808]: Users will need to explicitly specify `LimitedSwap` in this case, so failing cleanly may be a better experience

Comment (kannon92) [reply cmd/kubelet/app/server.go:808]: I think this PR is minutes away from merging so can i do a minor follow up?

Comment (liggitt) [reply cmd/kubelet/app/server.go:808]: follow up to fail fast at startup when explicitly configured to use limited swap on a machine where it cannot sgtm

Comment (mrunalp) [reply cmd/kubelet/app/server.go:808]: follow up sounds good :+1:

Comment (kannon92) [reply cmd/kubelet/app/server.go:808]: This PR is taking its sweet time to merge..  

Its like watching water boil.

I have https://github.com/kubernetes/kubernetes/pull/123738 as a quick follow up once this merges. I'll check and rebase accordingly when this merges.

Comment (yujuhong) [reply cmd/kubelet/app/server.go:808]: SG. Thanks

Comment (kannon92) [reply cmd/kubelet/app/server.go:808]: https://github.com/kubernetes/kubernetes/pull/123738 is up now and rebased.

**Pull request merged**
